drivers/platform/stm32/stm32_dma.c

```diff
@@ -0,0 +1,284 @@
+ * @param param - Initialization parameter for the decriptor.
+ * @return 0 in case of success
+ * 	   -ENOMEM if there is not enough free memory
+ */
+int stm32_dma_init(struct no_os_dma_desc** desc,
+		   const struct no_os_dma_init_param* param)
+{
+
+	struct no_os_dma_desc* descriptor;
+	int ret;
+	uint8_t i;
+
+	if (!desc)
+		return -EINVAL;
+
+	descriptor = no_os_calloc(1, sizeof(*descriptor));
+	if (!descriptor)
+		return -ENOMEM;
+
+	descriptor->channels = no_os_calloc(param->num_ch,
+					    sizeof(*descriptor->channels));
+	if (!descriptor->channels) {
+		ret = -ENOMEM;
+		goto free_descriptor;
+	}
+
+	descriptor->id = param->id;
+	descriptor->num_ch = param->num_ch;
+
+	struct no_os_irq_init_param irq_param = {
+		.irq_ctrl_id = 0,
+		.platform_ops = &stm32_irq_ops,
+		.extra = NULL
+	};
+
+	for (i = 0; i < param->num_ch; i++) {
+		descriptor->channels[i].free = true;
+	}
+
+	no_os_irq_ctrl_init(&descriptor->irq_ctrl, &irq_param);
+
+	*desc = descriptor;
+
+	return 0;
+
+free_descriptor:
+	no_os_free(descriptor);
+	descriptor = NULL;
+
+	return ret;
+}
+
+/**
+ * @brief Start a DMA transfer for a specific channel .
+ * @param desc - Descriptor for the DMA controller.
+ * @param ch - The index of the channel.
+ * @return 0 in case of success, negative error codes otherwise
+ */
+int stm32_dma_xfer_start(struct no_os_dma_desc *desc,
+			 struct no_os_dma_ch *chan)
+{
+	int ret;
+	struct stm32_dma_channel* sdma_chan;
+
+	if (!desc || !chan || !chan->extra)
+		return -EINVAL;
+
+	sdma_chan = chan->extra;
+
+	if (sdma_chan->irq_num) {
+		ret = HAL_DMA_Start_IT(sdma_chan->hdma, (uint32_t)sdma_chan->src,
+				       (uint32_t )sdma_chan->dst, sdma_chan->length);
+		if (ret != HAL_OK)
+			return -EINVAL;
+	} else {
+		ret = HAL_DMA_Start(sdma_chan->hdma, (uint32_t )sdma_chan->src,
+				    (uint32_t )sdma_chan->dst, sdma_chan->length);
+		if (ret != HAL_OK)
+			return -EINVAL;
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Disable a DMA channel.
+ * @param desc - Descriptor for the DMA controller.
+ * @param ch - The index of the channel.
+ * @return 0 in case of success, negative error codes otherwise
+ */
+int stm32_dma_xfer_abort(struct no_os_dma_desc* desc,
+			 struct no_os_dma_ch* chan)
+{
+	int ret;
+	struct stm32_dma_channel* sdma_chan;
+
+	if (!desc || !chan || !chan->extra)
+		return -EINVAL;
+
+	sdma_chan = chan->extra;
+
+	if (sdma_chan->irq_num) {
+		ret = HAL_DMA_Abort_IT(sdma_chan->hdma);
+		if (ret != HAL_OK)
+			return -EINVAL;
+	} else {
+		ret = HAL_DMA_Abort(sdma_chan->hdma);
+		if (ret != HAL_OK)
+			return -EINVAL;
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Free the resources allocated for a DMA descriptor.
+ * @param desc - Descriptor to be freed.
+ * @return 0 in case of success, negative error code otherwise
+ */
+int stm32_dma_remove(struct no_os_dma_desc* desc)
+{
+	int ret;
+	uint8_t i;
+
+	no_os_irq_ctrl_remove(desc->irq_ctrl);
+
+	for (i = 0; i < desc->num_ch; i++)
+		no_os_free(&desc->channels[i]);
```

Comment:
This is incorrect since you'll end up double freeing some memory. Should be `no_os_free(desc->channels)`

Reply:
I've removed free of channels. I suppose freeing up the desc should take care of this as well?

Reply:
You still need to free `desc->channels`. What I was referring to is to replace

```
for (i = 0; i < desc->num_ch; i++)
		no_os_free(&desc->channels[i]);
```

with

```
no_os_free(desc->channels);
```